### PR TITLE
Make userspace page cache work with 'web' disks

### LINKED ISSUE #61474: Userspace page cache doesn't work with disk `web`
See #60195. Stateful tests use disk `web` and most of them are failing with settings `enable_filesystem_cache = 0, use_page_cache_for_disks_without_file_cache = 1`.

There are at least two problems I found:
1. `getFileSize` is used by userspace page cache but it's not supported by `ReadBufferFromWebServer` (fix e51f084cbe3087d41a7d85b569d4bd4eea911ab8)
2. Userspace page cache uses seeks after creation of buffer (e.g. S3 read buffer is created with option `restricted_seek = false`) but it's not implemented in `ReadBufferFromWebServer`.

...Maybe something more but at that moment I gave up to fix it by myself and created this issue.

### PR DESCRIPTION
**Pull request by al13n321**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Userspace page cache works with static web storage (`disk(type = web)`) now. Use client setting `use_page_cache_for_disks_without_file_cache=1` to enable.

Closes https://github.com/ClickHouse/ClickHouse/issues/61474

```
:) select sum(ignore(*)) from uk_price_paid settings use_page_cache_for_disks_without_file_cache=0

SELECT sum(ignore(*))
FROM uk_price_paid
SETTINGS use_page_cache_for_disks_without_file_cache = 0

Query id: 8f22c410-6627-406b-9185-a3e5aba2030e

┌─sum(ignore(price, date, postcode1, postcode2, type, is_new, duration, addr1, addr2, street, locality, town, district, county))─┐
│                                                                                                                              0 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.972 sec. Processed 27.91 million rows, 1.24 GB (28.72 million rows/s., 1.27 GB/s.)
Peak memory usage: 87.18 MiB.

:) select sum(ignore(*)) from uk_price_paid settings use_page_cache_for_disks_without_file_cache=1

SELECT sum(ignore(*))
FROM uk_price_paid
SETTINGS use_page_cache_for_disks_without_file_cache = 1

Query id: abb429c8-6fdb-4939-9b67-92226ac2877c

┌─sum(ignore(price, date, postcode1, postcode2, type, is_new, duration, addr1, addr2, street, locality, town, district, county))─┐
│                                                                                                                              0 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.688 sec. Processed 27.91 million rows, 1.24 GB (40.56 million rows/s., 1.80 GB/s.)
Peak memory usage: 148.41 MiB.

:) select sum(ignore(*)) from uk_price_paid settings use_page_cache_for_disks_without_file_cache=1

SELECT sum(ignore(*))
FROM uk_price_paid
SETTINGS use_page_cache_for_disks_without_file_cache = 1

Query id: a21093c7-f802-4131-b3ab-358971b4c6bb

┌─sum(ignore(price, date, postcode1, postcode2, type, is_new, duration, addr1, addr2, street, locality, town, district, county))─┐
│                                                                                                                              0 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.218 sec. Processed 27.91 million rows, 1.24 GB (128.25 million rows/s., 5.68 GB/s.)
Peak memory usage: 199.04 MiB.
```

The third query made no HTTP requests.

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit d0835c11ae8fad70d6d4cf521d68859658c3f6ee with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61911/d0835c11ae8fad70d6d4cf521d68859658c3f6ee/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by al13n321** [comment]
Now even cold cache makes it faster than no cache, because a query reads the same data multiple times for some reason. For the example table from https://clickhouse.com/docs/en/operations/storing-data#web-storage , `select sum(ignore(*)) from uk_price_paid` reads 230 MB without cache, 177 MB with cold cache; the table is 177 MB on disk.

**Comment by al13n321** [comment]
Failed tests are unrelated:
* 01193_metadata_loading - https://github.com/ClickHouse/ClickHouse/pull/61961
* AST Fuzzer (ubsan) - https://github.com/ClickHouse/ClickHouse/issues/61962

**Comment by al13n321** [comment]
Other failed tests are also unrelated:
* [Integration Tests Flaky Check (asan)](https://s3.amazonaws.com/clickhouse-test-reports/61911/1b828ab6f60574317853378262f78754ca7849d8/integration_tests_flaky_check__asan_.html) - timed out waiting for download from dockerhub (if I'm reading the code correctly).
* [performance comparison](https://s3.amazonaws.com/clickhouse-test-reports/61911/1b828ab6f60574317853378262f78754ca7849d8/performance_comparison_[1_4]/report.html) - 1 out of 396 queries got p-value 0.01. Why only 1? Shouldn't 3.96 queries get p-value < 0.01, on average? Or does the "p < 0.01 threshold" already include Bonferroni correction? The query looks unrelated to the changes in this PR.

File: tests/integration/test_disk_over_web_server/test.py [hunk 1]
```diff
@@ -322,3 +322,74 @@ def test_replicated_database(cluster):
...
+        assert (
+            ev1.get("DiskConnectionsCreated", 0) + ev1.get("DiskConnectionsReused", 0)
+            > 0
+        )
+
+        ev2 = get_profile_events("test warm cache")
+        assert ev2.get("PageCacheChunkDataHits", 0) > 0
+        assert ev2.get("PageCacheChunkMisses", 0) == 0
+        assert (
+            ev2.get("DiskConnectionsCreated", 0) + ev2.get("DiskConnectionsReused", 0)
```
Comment (azat) [inline tests/integration/test_disk_over_web_server/test.py:382]: @al13n321 this check fails from time to time, can you please take a look?
https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSB0ZXN0X25hbWUgTElLRSAnJXRlc3RfcGFnZV9jYWNoZSUnCiAgICBBTkQgY2hlY2tfc3RhcnRfdGltZSA+PSBub3coKSAtIElOVEVSVkFMIDcgZGF5Ci8vICAgIEFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gMAogICAgQU5EIHRlc3Rfc3RhdHVzICE9ICdTS0lQUEVEJwogICAgQU5EIHRlc3Rfc3RhdHVzIExJS0UgJ0YlJwogICAgQU5EIGNoZWNrX3N0YXR1cyAhPSAnc3VjY2VzcycKT1JERVIgQlkgY2hlY2tfbmFtZSwgdGVzdF9uYW1lLCBjaGVja19zdGFydF90aW1l

**Pull request merged**